### PR TITLE
abis: Define struct msqid_ds

### PR DESCRIPTION
**Pull request by Dennisbonke**
Part of the mlibc LFS project.

### CODE REVIEW

File: options/posix/include/sys/msg.h [hunk 1]
```diff
@@ -2,13 +2,32 @@
...
+	time_t msg_stime;
+	time_t msg_rtime;
+	time_t msg_ctime;
+	unsigned long __msg_cbytes;
+	msgqnum_t msg_qnum;
+	msglen_t msg_qbytes;
+	pid_t msg_lspid;
+	pid_t msg_lrpid;
+	unsigned long __glibc_reserved4;
+	unsigned long __glibc_reserved5;
```
Comment (ArsenArsen) [inline options/posix/include/sys/msg.h:28]: Is this in violation of the GPL?

**Review by ArsenArsen — CHANGES_REQUESTED** [review]
Is this correct? Also, isn't this an abi-bit?

File: options/posix/include/sys/msg.h (hunk 1 above)
Comment (Dennisbonke) [reply options/posix/include/sys/msg.h:28]: Can you copyright struct definitions? No clue, but if it's a problem, we can rename them.

**Comment by Dennisbonke** [comment]
> Is this correct? Also, isn't this an abi-bit?

Don't see why it isn't correct, and the distinction between needs to be abi-bit and can go into the main header remains unclear for me at times. If the mlibc abi matches the linux abi (aka here) then why do the extra work?

**Comment by ArsenArsen** [comment]
Because this doesn't seem to match between Linux architectures

File: options/posix/include/sys/msg.h (hunk 1 above)
Comment (Dennisbonke) [reply options/posix/include/sys/msg.h:28]: Well that's gonna be fun if I need struct definitions and musl doesn't have them. Can't use different names either as programs rely on them (and no, they aren't standard per se). I see no problems copying definitions, however, copying function implementations is indeed not done.

Comment (Geertiebear) [reply options/posix/include/sys/msg.h:28]: Why are the reserved fields copied over anyways?

File: options/posix/include/sys/msg.h [hunk 2]
```diff
@@ -2,13 +2,32 @@
...
+typedef unsigned long msgqnum_t;
+
 int msgget(key_t, int);
 
+struct msqid_ds {
+	struct ipc_perm msg_perm;
+	time_t msg_stime;
+	time_t msg_rtime;
+	time_t msg_ctime;
+	unsigned long __msg_cbytes;
```
Comment (avdgrinten) [inline options/posix/include/sys/msg.h:22]: This field is called `msg_cbytes` according to the man page.

File: options/posix/include/sys/msg.h (hunk 1 above)
Comment (avdgrinten) [reply options/posix/include/sys/msg.h:28]: Just rename them to `unsigned long __unused[2]` as musl does.

File: abis/linux/msg.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <sys/ipc.h>
+#include <bits/ansi/time_t.h>
+#include <abi-bits/pid_t.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef unsigned long msglen_t;
+typedef unsigned long msgqnum_t;
```
Comment (ArsenArsen) [inline abis/linux/msg.h:13]: should we make these fixed-width? how do they differ across the arches we have?

**Review by ArsenArsen — COMMENTED** [review]
`typedef __syscall_ulong_t msgqnum_t;` is the glibc definition for this. does __syscall_ulong_t always match unsigned long?

File: abis/linux/msg.h (hunk 3 above)
Comment (ArsenArsen) [reply abis/linux/msg.h:13]: okay, I looked around the logic here a bit, I think these are always `unsigned long int`, but they depend on `__WORDSIZE`, and I'm not sure how we handle that case

Comment (Geertiebear) [reply abis/linux/msg.h:13]: We can copy the logic that depends on __WORDSIZE in a static assert.

**Comment by Geertiebear** [comment]
This PR currently has merge conflicts.

**Comment by Dennisbonke** [comment]
I've fixed the conflicts, can this get another review?

**Comment by no92** [comment]
Rebased onto current master, dropping vinix and adding lyre.

File: abis/linux/msg.h (hunk 3 above)
Comment (Geertiebear) [reply abis/linux/msg.h:13]: This comment is still not addressed, maybe @ArsenArsen can link to the logic in question so it's easier to copy over?

Comment (no92) [reply abis/linux/msg.h:13]: Looking through glibc behavior, these are always defined to be `unsigned long int`, except for i386, where they are `uint64_t`.

**Comment by no92** [comment]
Fixed the typedefs for `msglen_t` and `msgqnum_t`.

**Review by Geertiebear — APPROVED** [review]
LGTM.

**Pull request merged**